Filemanager continuous scrolling

lhietal opened the pull request:
Continues  #8408.
Allow thumbs to not align at the top of filemanager with `GDK_SCROLL_SMOOTH` events. Provides nicer scrolling experience on a touchpad. Scrollwheel behaviour should not have changed. Tested on macOS 10.14 with touchpad and traditional mouse. I'm unable to test how this works on other OSs and I think smooth scrolling can be handled very differently.

There are a few issues that might need adressing. 

- I have also been experimenting with enabling smooth scrolling for other parts of the UI and scaling by 100 seems appropriate. Does it work on other OSs? 
- When scrolling very fast it is possible to scroll past top row.
- With large amount of thumbs visible scrolling can be slow and not smooth. Previously commented #2226.
- Iops that are also using `dt_gui_get_scroll_deltas()` seem to work as before despite scaling.

It would be nice to hear how this change affects other systems. I'm reverting changes made in #5503 so I'm worried that other systems might not work properly.

TurboGit commented:
Not sure about this. Fact is that when scrolling it is now hard to have no cut images top or bottom. So one need to use the scrollbar to align on top. I'll let @AlicVB comment on this as he does the rework of the lighttable. For me we would at least need some kind of auto alignment on top as a magnetic-grid or something like that to ease having full top line visible.

AlicVB reviewed (changes requested):
I fear that this is not what we want :(
Here (on linux with x11) this has the following effects : 
1. scroll with the mouse ("normal" mouse) make scrolling way slower as one wheel "step" only move for a subset of the thumb.
2. image are never aligned on top (as @TurboGit point)
3. as movements are "hardcoded" to 100 (see further) the scroll is no longer tight to thumb size...
4. I fear that it break things when used with trackpads...

tbh, I'm not sure to understand the root issue it solves, at least on Linux (sadly I've no access to machine with other OS)

lhietal commented:
I'm only talking about scrolling with a touchpad (might apply to touchscreens as well). Touchpad doesn't have any haptic feedback for when there is a change in the UI. Therefore I think it is better to have the content act as if it's "attached to my fingers". There is a corporally understandable connection between the hand movement and filemanager thumbnails. Analogous to e.g. sliding paper on a table.

When scrolling through thumbnails I am primarily looking at 1/3 from the top of the screen. Having images align on top is not crucial. I'm pretty sure that touchpad experience on Linux is different so we might be talking about different things.

Behaviour change with normal mouse is not intended. Perhaps some mice also send smooth scroll events.

AlicVB commented:
> I'm only talking about scrolling with a touchpad (might apply to touchscreens as well). Touchpad doesn't have any haptic feedback for when there is a change in the UI. Therefore I think it is better to have the content act as if it's "attached to my fingers". There is a corporally understandable connection between the hand movement and filemanager thumbnails. Analogous to e.g. sliding paper on a table.

ok thanks for the explanation, it's more clear :)



> Having images align on top is not crucial

agreed for touchpads, not for "regular" mouse with "physical" step on the wheel



> Behaviour change with normal mouse is not intended. Perhaps some mice also send smooth scroll events.

At least mine is doing that (logitech MX master 2S) whatever the mode of the wheel (with steps or not)

In fact, I think that the main point is to restrict the new behavior to touchpads only. The question is how to know that, considering that e.direction doesn't seems reliable enough ?

elstoc commented:
I use a touchpad and I much prefer the images to snap to the top of the screen (i.e. for touchpad scrolling to behave like mouse scrolling).

parafin commented on src/dtgtk/thumbtable.c:
```diff
@@ -878,8 +890,30 @@ static gboolean _event_scroll(GtkWidget *widget, GdkEvent *event, gpointer user_
+    {
+      // Scale delta to represent scroll size in pixels.
+      // On macOS reverses scaling in dt_gui_get_scroll_deltas
+      delta = delta * 100;
```
And on other OSes what does this do?..

lhietal replied:
I haven't tested other OSes. Scaling in _dt_gui_get_scroll_deltas()_ makes macOS work with other controls (not part of this PR). So I assumed that made all OSes work similarly. Definitely needs testing on other systems.

I think using scroll deltas directly is how scrolling is intended to work. That is why I added the comment.

lhietal commented:
It seems like there is a way to get the type of device. There is a similar situation in gtk.c line 1165 where graphics tablet is handeled separately. It sould be possible to get touchpad as well. ([gdk3](https://developer.gnome.org/gdk3/stable/GdkDevice.html#GdkInputSource)) But I'm not able to get it working. Both touchpad and mouse events come from _Core pointer_ which is _GDK_SOURCE_MOUSE_.

> I use a touchpad and I much prefer the images to snap to the top of the screen (i.e. for touchpad scrolling to behave like mouse scrolling), though I don't know how it would be impacted by this PR.

I'm not sure how to proceed with this. There has been at least one other person who [prefers continuous scrolling](https://github.com/darktable-org/darktable/issues/6068#issuecomment-774664695). Do we want a setting for this behaviour?

lhietal commented:
I implemented clamping for continuous scrolling and added an entry in preferences. I haven't been able to come up with anything better to allow normal mice to work as before. I'm not sure how this works on Linux or Windows with a touchpad.

Possible problems with **continuous scrolling enabled**
- normal scrollwheel aligns images at the bottom edge after scrolling to bottom
- normal mouse isn't always detected (sends smooth scroll events) so scroll amount is incorrect

lhietal commented:
I tested with Ubuntu 20.10 and Macbook touchpad. It works fine. I'm not able to test Windows.

I have no ideas to improve the behaviour further. I'll make another PR with more smooth scrolling improvements on MacOS.

lhietal commented:
Reset offset only when collection changes. Otherwise it is difficult to set ratings and labels. Scaling is now same as in #8616. Doesn't affect macOS but my testing on Ubuntu 20.10 shows slightly slower scrolling. For me it feels better now.

lhietal commented:
Scrollbars are now also working without steps. I changed scrollbar dragging to not align thumbs (done in #8408) only when continuous scrolling is enabled. Moving with arrow keys wasn't working correctly. This is now fixed as well as some improvements in clamping.

lhietal commented:
I think this is finished now. Next step could be to implement similar change for filmstrip as well but I don't want to start that unless I know that this PR can be merged.

> In fact, I think that the main point is to restrict the new behavior to touchpads only. The question is how to know that, considering that e.direction doesn't seems reliable enough ?

@AlicVB Currently there is an entry in the preferences to disable continuous scrolling. I tried to restrict the change to touchpads on top of this but couldn't find a solution. Perhaps it would be better to handle mixed smooth and regular scroll events in gtk.c because continuous scrolling could also be wanted with a mouse. https://github.com/darktable-org/darktable/blob/40a5e9cc3e321efb5e9ace5479ffb8390033807f/src/gui/gtk.c#L553

I can't reproduce the mixed events behaviour with my hardware so it is difficult to make concrete suggestions.

There is a special case with moving thumbtable with arrow keys where I'm not sure what would be the best behaviour. It's not always possible to get last thumb row visible by just pressing down arrow. If last row isn't completely filled last full row can align at the bottom edge and moving further isn't possible unless there is a thumb below currently active one. This could be changed by making the last thumb in collection active when trying to go past last thumb of any column.

TurboGit commented:
@lhietal : Sorry for the long delay, what is the status of this PR ?

TurboGit commented:
@lhietal : Is the image cut is now avoided (see my comment above). This was my main complaints, and I won't see introducing a usability issue for a continuous scroll has improvement.

lhietal commented:
Closing seems reasonable given that nobody else has shown interest in this feature.
